blog/tanstack-router-loaders/index.md

```diff
@@ -0,0 +1,630 @@
+}
+```
+
+Two queries on this page: one to get the list of (paged) epics, another to get the total count of all the epics. Let's run it
+
+![loaders running in parallel](/tanstack-router-loaders/img-2-epics-rendered.jpg)
+
+It's as silly as before, but it does show the three pieces of data we've fetched: the overview data we fetched in the epics layout; and then the count of epics, and the list of epics we loaded in the epics page, beneath that.
+
+What's more, when we run this, we first see the the pending component for our root route. That resolves quickly, and shows the main navigation, along with the pending component for our epics route. That resolves, showing the epics overview data, and then revealing the pending component for our epics page, which eventually resolves and shows the list, and count of our epics.
+
+Our component-level data fetching is working, and integrating, via Suspense, with the same Router pending components we already had. Very cool!
+
+Let's take a peak at our console though, and look at all the various logging we've been doing, to track when these fetches happen
+
+![loaders running in parallel](/tanstack-router-loaders/img-3-epics-waterfall.jpg)
+
+The results are ... awful. Component-level data fetching with Suspense feels really good, but if you're not careful, these waterfalls are extremely easy to create. The problem is, when a component suspends while waiting for data, it prevents its children from rendering. This is precisely what's happening here. The route is suspending, and not even giving the child component, which includes the page (and any other nested route components underneath) from rendering, which prevents those components' fetches from starting.
+
+There's two potential solutions here: we could dump Suspense, and use the `useQuery` hook, instead, which does not suspend. That would require us to manually track multiple isLoading states (for each useQuery hook), and coordinate loading UX to go with that. For the epics page, we'd need to track both the count loading state, and the epics list state, and not show our UI until both have returned. And so on, for every other page.
+
+The other solution is to start pre-fetching these queries sooner.
+
+We'll go with option 2.
+
+### Prefetching
+
+Remember previously we saw that loader functions all run in parallel. This is the perfect opportunity to start these queries off ahead of time, before the components even render. TanStack Query gives us an api to do just that.
+
+To prefetch with Query, we take the `queryClient` object we saw before, and call `queryClient.prefetchQuery` and pass in **the exact same query options** and Query will be smart enough, when the component loads and executes `useSuspenseQuery`, to see that the query is already in flight, and just latch onto that same request. That's also a big reason why we put those query options into the `epicsSummaryQueryOptions` helper function: to make it easier to reuse in the loader, to prefetch.
```

Comment:
we tend to recommend `ensureQueryDate` over `prefetchQuery`

Reply:
hmm eager to know more. They seem like different api's designed for different things